File: configuration/src/test/java/com/synopsys/integration/configuration/property/types/enums/EnumPropertiesTests.java
package com.synopsys.integration.configuration.property.types.enums;

import static com.synopsys.integration.configuration.util.ConfigTestUtils.configOf;
import static com.synopsys.integration.configuration.util.ConfigTestUtils.emptyConfig;

import java.util.List;
import java.util.Optional;

import org.apache.commons.lang3.tuple.Pair;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import com.synopsys.integration.common.util.Bds;
import com.synopsys.integration.configuration.config.InvalidPropertyException;
import com.synopsys.integration.configuration.config.PropertyConfiguration;
import com.synopsys.integration.configuration.property.PropertyTestHelpUtil;
import com.synopsys.integration.configuration.property.deprecation.DeprecatedValueUsage;

// Simple glue sanity tests. Theoretically if Config is well tested and Parser is well tested, these will pass so they are not exhaustive.
public class EnumPropertiesTests {
    private enum Example {
        THING,
        ANOTHER,
        THIRD
    }

    @Test
    public void testExampleValues() {
        NullableEnumProperty<Example> property = new NullableEnumProperty<>("example.list", Example.class);
        Assertions.assertEquals(Bds.listOf("THING", "ANOTHER", "THIRD"), property.listExampleValues());
    }

    @Test
    public void testNullable() throws InvalidPropertyException {
        NullableEnumProperty<Example> property = new NullableEnumProperty<>("enum.nullable", Example.class);
        PropertyConfiguration config = configOf(Pair.of("enum.nullable", "ANOTHER"));
        Assertions.assertEquals(Optional.of(Example.ANOTHER), config.getValue(property));

        PropertyTestHelpUtil.assertAllHelpValid(property, Bds.listOf("THING", "ANOTHER", "THIRD"));
    }

    @Test
    public void testValued() throws InvalidPropertyException {
        EnumProperty<Example> property = new EnumProperty<>("enum.valued", Example.THIRD, Example.class);
        PropertyConfiguration config = configOf(Pair.of("enum.valued", "THIRD"));
        Assertions.assertEquals(Example.THIRD, config.getValue(property));

        PropertyTestHelpUtil.assertAllHelpValid(property, Bds.listOf("THING", "ANOTHER", "THIRD"));
    }

    @Test
    public void testDeprecatedValue() throws InvalidPropertyException {
        EnumProperty<Example> property = new EnumProperty<>("enum.valued", Example.ANOTHER, Example.class);
        property.deprecateValue(Example.THIRD, "Third is deprecated");

        PropertyConfiguration config = configOf(Pair.of("enum.valued", "THIRD"));
        Optional<Example> value = config.getProvidedParsedValue(property);
        Assertions.assertTrue(value.isPresent());

        List<DeprecatedValueUsage> deprecatedUsages = property.checkForDeprecatedValues(value.get());
        Assertions.assertEquals(deprecatedUsages.size(), 1);
        Assertions.assertEquals(deprecatedUsages.get(0).getValue(), "THIRD");
        Assertions.assertEquals(deprecatedUsages.get(0).getInfo().getValueDescription(), "THIRD");
        Assertions.assertEquals(deprecatedUsages.get(0).getInfo().getReason(), "Third is deprecated");
    }

    @Test
    public void testDeprecatedValueDefaultNotUsage() throws InvalidPropertyException {
        EnumProperty<Example> property = new EnumProperty<>("enum.valued", Example.THIRD, Example.class);
        property.deprecateValue(Example.THIRD, "Third is deprecated");

        PropertyConfiguration config = emptyConfig();
        Optional<Example> value = config.getProvidedParsedValue(property);
        Assertions.assertFalse(value.isPresent());
    }

    @Test
    public void testDeprecatedValueDefaultProvided() throws InvalidPropertyException {
        EnumProperty<Example> property = new EnumProperty<>("enum.valued", Example.THIRD, Example.class);
        property.deprecateValue(Example.THIRD, "Third is deprecated");

        PropertyConfiguration config = configOf(Pair.of("enum.valued", "THIRD"));
        Optional<Example> value = config.getProvidedParsedValue(property);
        Assertions.assertTrue(value.isPresent());

        List<DeprecatedValueUsage> deprecatedUsages = property.checkForDeprecatedValues(value.get());
        Assertions.assertEquals(deprecatedUsages.size(), 1);
        Assertions.assertEquals(deprecatedUsages.get(0).getValue(), "THIRD");
        Assertions.assertEquals(deprecatedUsages.get(0).getInfo().getValueDescription(), "THIRD");
        Assertions.assertEquals(deprecatedUsages.get(0).getInfo().getReason(), "Third is deprecated");
    }

    @Test
    public void testList() throws InvalidPropertyException {
        EnumListProperty<Example> property = new EnumListProperty<>("enum.list", Bds.listOf(Example.THIRD), Example.class);
        PropertyConfiguration config = configOf(Pair.of("enum.list", "ANOTHER,THING"));
        Assertions.assertEquals(Bds.listOf(Example.ANOTHER, Example.THING), config.getValue(property));

        PropertyTestHelpUtil.assertAllListHelpValid(property);
    }
}

File: configuration/src/test/java/com/synopsys/integration/configuration/property/types/enum/EnumPropertiesTests.kt
/**
 * configuration
 *
 * Copyright (c) 2020 Synopsys, Inc.
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package com.synopsys.integration.configuration.property.types.enum

import com.synopsys.integration.configuration.property.PropertyTestHelpUtil
import com.synopsys.integration.configuration.property.types.enums.EnumListProperty
import com.synopsys.integration.configuration.property.types.enums.EnumProperty
import com.synopsys.integration.configuration.property.types.enums.NullableEnumProperty
import com.synopsys.integration.configuration.util.ConfigTestUtils.configOf
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.util.*

// Simple glue sanity tests. Theoretically if Config is well tested and Parser is well tested, these will pass so they are not exhaustive.
class EnumPropertiesTests {
    private enum class Example {
        THING,
        ANOTHER,
        THIRD
    }

    @Test
    fun testExampleValues() {
        val property = NullableEnumProperty("example.list", Example::class.java)
        Assertions.assertEquals(listOf("THING", "ANOTHER", "THIRD"), property.listExampleValues());
    }

    @Test
    fun testNullable() {
        val property = NullableEnumProperty("enum.nullable", Example::class.java)
        val config = configOf("enum.nullable" to "ANOTHER")
        Assertions.assertEquals(Optional.of(Example.ANOTHER), config.getValue(property))

        PropertyTestHelpUtil.assertAllHelpValid(property, expectedExampleValues = listOf("THING", "ANOTHER", "THIRD"))
    }

    @Test
    fun testValued() {
        val property = EnumProperty("enum.valued", Example.THIRD, Example::class.java)
        val config = configOf("enum.valued" to "THIRD")
        Assertions.assertEquals(Example.THIRD, config.getValue(property))

        PropertyTestHelpUtil.assertAllHelpValid(property, expectedExampleValues = listOf("THING", "ANOTHER", "THIRD"))
    }

    @Test
    fun testList() {
        val property = EnumListProperty("enum.list", listOf(Example.THIRD), Example::class.java)
        val config = configOf("enum.list" to "ANOTHER,THING")
        Assertions.assertEquals(listOf(Example.ANOTHER, Example.THING), config.getValue(property))

        PropertyTestHelpUtil.assertAllHelpValid(property)
    }
}